gen/generator.jl

```diff
@@ -0,0 +1,98 @@
+using Clang, Clang.LibClang
+using CEnum
+using DataStructures, OrderedCollections 
+
+include("clanghelpers.jl")
+
+const HSA_INCLUDE = joinpath("..", "deps", "include", "hsa") |> normpath
+const HSA_HEADERS = [joinpath(HSA_INCLUDE, header) for header in readdir(HSA_INCLUDE) if endswith(header, ".h")]
+
+ctx = DefaultContext()
+
+parse_headers!(ctx, HSA_HEADERS,
+               args=["-I", joinpath(HSA_INCLUDE, ".."), "-fparse-all-comments"],
+               includes=vcat(HSA_INCLUDE, CLANG_INCLUDE),
+               flags = CXTranslationUnit_DetailedPreprocessingRecord
+               )
+
+ctx.libname = "libhsa"
+ctx.options["is_function_strictly_typed"] = true
+ctx.options["is_struct_mutable"] = false
+    
+blacklist = ["unix", "linux"]
+blackprefix = ["Brig", "_"]
+
+functions = Vector{CLCursor}()
+remaining = Vector{CLCursor}()
+typealiases = OrderedDict()
+
+for trans_unit in ctx.trans_units
+
+    root_cursor = getcursor(trans_unit)
+    push!(ctx.cursor_stack, root_cursor)
+    header = spelling(root_cursor)
+    ctx.children = children(root_cursor)
+
+    for (i, child) in enumerate(ctx.children)
+        # Cursor properties
+        child_name = name(child)
+        child_kind = kind(child)
+        if i > 1
+            prev_name = name(ctx.children[i-1])
+            prev_kind = kind(ctx.children[i-1])
+        end
+        if i < length(ctx.children)
+            next_name = name(ctx.children[i+1])
+            next_kind = kind(ctx.children[i+1])
+        end
+        ctx.children_index = i
+        
+        # Skip compiler constants/garbage and redundants
+        child_kind == CXCursor_MacroInstantiation   && continue
+        child_kind == CXCursor_LastPreprocessing    && continue
+        child_name ∈ blacklist                      && continue
+        any(startswith.(child_name, blackprefix))   && continue
+        child_name ∈ keys(ctx.common_buffer)        && continue
+        child_name ∈ name.(functions)               && continue 
+        child_kind == CXCursor_FunctionDecl     && (wrap_hsa_fun!(ctx, child); push!(functions, child); continue)
+        child_kind == CXCursor_StructDecl       && (wrap_hsa_struct!(ctx, child); continue) 
+        child_kind == CXCursor_EnumDecl         && (wrap!(ctx, child); continue)
+        child_kind == CXCursor_MacroDefinition  && (wrap!(ctx, child); continue)
+        child_kind == CXCursor_VarDecl          && (wrap!(ctx, child); continue)
+
+#=        # skip redundant and most unneeded types
```

Comment:
Is this useful to keep around for the near future, or is it something we can strip out?

Reply:
Not 100% no. I think I wound up giving up trying to suppress the separate type definitions of the associated struct and enum tags and just deleted them from the output manually. Unfortunately I never got all the way to full automation--the generator script does maybe 80% of the work. I left things like that in there because perhaps later (with renewed energy and a better version of Clang.jl, they might prove useful toward getting _actual_ 100% automated wrappers.